File: typescript/src/key-pair.ts
import {
  autograph_session_key_pair,
  autograph_session_key_pair_size,
  autograph_identity_key_pair,
  autograph_identity_key_pair_size,
  autograph_get_identity_public_key,
  autograph_get_session_public_key,
  autograph_identity_public_key_size,
  autograph_session_public_key_size
} from './clib'

export const createIdentityKeyPair = () =>
  new Uint8Array(autograph_identity_key_pair_size())

export const createSessionKeyPair = () =>
  new Uint8Array(autograph_session_key_pair_size())

export const createIdentityPublicKey = () =>
  new Uint8Array(autograph_identity_public_key_size())

export const createSessionPublicKey = () =>
  new Uint8Array(autograph_session_public_key_size())

export const generateIdentityKeyPair = (): Uint8Array => {
  const keyPair = createIdentityKeyPair()
  const success = autograph_identity_key_pair(keyPair)
  if (!success) {
    throw new Error('Identity key pair generation failed')
  }
  return keyPair
}

export const generateSessionKeyPair = (): Uint8Array => {
  const keyPair = createSessionKeyPair()
  const success = autograph_session_key_pair(keyPair)
  if (!success) {
    throw new Error('Session key pair generation failed')
  }
  return keyPair
}

export const getIdentityPublicKey = (keyPair: Uint8Array) => {
  const publicKey = createIdentityPublicKey()
  autograph_get_identity_public_key(publicKey, keyPair)
  return publicKey
}

export const getSessionPublicKey = (keyPair: Uint8Array) => {
  const publicKey = createSessionPublicKey()
  autograph_get_session_public_key(publicKey, keyPair)
  return publicKey
}

export const getPublicKeys = (
  identityKeyPair: Uint8Array,
  sessionKeyPair: Uint8Array
): [Uint8Array, Uint8Array] => {
  const identityKey = getIdentityPublicKey(identityKeyPair)
  const sessionKey = getSessionPublicKey(sessionKeyPair)
  return [identityKey, sessionKey]
}

File: rust/src/key_pair.rs
use alloc::{vec, vec::Vec};
use rand_core::{CryptoRng, RngCore};

use crate::{
    error::Error,
    primitives::{DiffieHellmanPrimitive, SigningPrimitive},
};

pub fn create_identity_key_pair<P: SigningPrimitive>() -> Vec<u8> {
    let size = P::IDENTITY_PRIVATE_KEY_SIZE + P::IDENTITY_PUBLIC_KEY_SIZE;
    vec![0; size]
}

pub fn create_session_key_pair<P: DiffieHellmanPrimitive>() -> Vec<u8> {
    let size = P::SESSION_PRIVATE_KEY_SIZE + P::SESSION_PUBLIC_KEY_SIZE;
    vec![0; size]
}

pub fn create_identity_public_key<P: SigningPrimitive>() -> Vec<u8> {
    vec![0; P::IDENTITY_PUBLIC_KEY_SIZE]
}

pub fn create_session_public_key<P: DiffieHellmanPrimitive>() -> Vec<u8> {
    vec![0; P::SESSION_PUBLIC_KEY_SIZE]
}

pub fn generate_session_key_pair<T: RngCore + CryptoRng, P: DiffieHellmanPrimitive>(
    csprng: T,
) -> Result<Vec<u8>, Error> {
    let mut key_pair = create_session_key_pair::<P>();
    if P::key_pair_session(csprng, &mut key_pair) {
        Ok(key_pair)
    } else {
        Err(Error::KeyPair)
    }
}

pub fn generate_identity_key_pair<T: RngCore + CryptoRng, P: SigningPrimitive>(
    csprng: T,
) -> Result<Vec<u8>, Error> {
    let mut key_pair = create_identity_key_pair::<P>();
    if P::key_pair_identity(csprng, &mut key_pair) {
        Ok(key_pair)
    } else {
        Err(Error::KeyPair)
    }
}

pub fn get_identity_public_key<P: SigningPrimitive>(key_pair: &[u8]) -> Vec<u8> {
    let mut public_key = create_identity_public_key::<P>();
    public_key.copy_from_slice(&key_pair[P::IDENTITY_PRIVATE_KEY_SIZE..]);
    public_key
}

pub fn get_session_public_key<P: DiffieHellmanPrimitive>(key_pair: &[u8]) -> Vec<u8> {
    let mut public_key = create_session_public_key::<P>();
    public_key.copy_from_slice(&key_pair[P::SESSION_PRIVATE_KEY_SIZE..]);
    public_key
}

pub fn get_public_keys<P: SigningPrimitive + DiffieHellmanPrimitive>(
    identity_key_pair: &[u8],
    session_key_pair: &[u8],
) -> (Vec<u8>, Vec<u8>) {
    let identity_key = get_identity_public_key::<P>(identity_key_pair);
    let session_key = get_session_public_key::<P>(session_key_pair);
    (identity_key, session_key)
}
